L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc

```diff
@@ -0,0 +1,134 @@
+                                                                                                  binsPhi_(conf.getParameter<std::vector<double> >("binsPhi")) {
+
+
+  if(!binsEta_.empty() && ((unsigned int)(binsEta_.size()) != nBinsEta_+1)) {
+    throw edm::Exception(edm::errors::Configuration, "Configuration")
+      << "HGCalTriggerTowerGeometryHelper nBinsEta for the tower map not consistent with binsEta size"<<std::endl;
+  }
+
+  if(!binsPhi_.empty() && ((unsigned int)(binsPhi_.size()) != nBinsPhi_+1)) {
+    throw edm::Exception(edm::errors::Configuration, "Configuration")
+      << "HGCalTriggerTowerGeometryHelper nBinsPhi for the tower map not consistent with binsPhi size"<<std::endl;
+  }
+
+  // if the bin vecctor is empty we assume the bins to be regularly spaced
+  if(binsEta_.empty()) {
+    for(unsigned int bin1 = 0; bin1 != nBinsEta_+1; bin1++) {
+      binsEta_.push_back(minEta_+bin1*((maxEta_-minEta_)/nBinsEta_));
+    }
+  }
+
+  // if the bin vecctor is empty we assume the bins to be regularly spaced
+  if(binsPhi_.empty()) {
+    for(unsigned int bin2 = 0; bin2 != nBinsPhi_+1; bin2++) {
+      binsPhi_.push_back(minPhi_+bin2*((maxPhi_-minPhi_)/nBinsPhi_));
+    }
+  }
+
+  for(int zside = -1; zside <= 1; zside+=2) {
+    for(unsigned int bin1 = 0; bin1 != nBinsEta_; bin1++) {
+      for(unsigned int bin2 = 0; bin2 != nBinsPhi_; bin2++) {
+        l1t::HGCalTowerID towerId(zside, bin1, bin2);
+        tower_coords_.emplace_back(towerId.rawId(),
+                                   (binsEta_[bin1+1] + binsEta_[bin1])/2,
+                                   (binsPhi_[bin2+1] + binsPhi_[bin2])/2);
+      }
+    }
+  }
+
+  if(conf.getParameter<bool>("readMappingFile")) {
+    // We read the TC to TT mapping from file,
+    // otherwise we derive the TC to TT mapping on the fly from eta-phi coord. of the TCs
+    std::ifstream l1tTriggerTowerMappingStream(conf.getParameter<edm::FileInPath>("L1TTriggerTowerMapping").fullPath());
+    if(!l1tTriggerTowerMappingStream.is_open()) {
+        throw cms::Exception("MissingDataFile")
+            << "Cannot open HGCalTriggerGeometry L1TTriggerTowerMapping file\n";
+    }
+
+    unsigned trigger_cell_id = 0;
+    unsigned short iEta = 0;
+    unsigned short iPhi = 0;
+
+    for(; l1tTriggerTowerMappingStream >> trigger_cell_id >> iEta >> iPhi;) {
+      if(iEta >= nBinsEta_ || iPhi >= nBinsPhi_) {
+        throw edm::Exception(edm::errors::Configuration, "Configuration")
+          << "HGCalTriggerTowerGeometryHelper warning inconsistent mapping TC : " << trigger_cell_id
+          << " to TT iEta: " << iEta << " iPhi: " << iPhi
+          << " when max #bins eta: "  << nBinsEta_ << " phi: " << nBinsPhi_ << std::endl;
+      }
+      HGCalDetId detId(trigger_cell_id);
+      int zside = detId.zside();
+      l1t::HGCalTowerID towerId(zside, iEta, iPhi);
```

Comment:
could avoid the temporary with `l1t::HGCalTowerID towerId(HGCalDetId(trigger_cell_id).zside(), iEta, iPhi);`

Reply:
however `HGCalDetId` won't work for the new geometry